include/psen_scan/async_udp_reader.h

```diff
@@ -0,0 +1,257 @@
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+#ifndef PSEN_SCAN_ASYNC_UDP_READER_H
+#define PSEN_SCAN_ASYNC_UDP_READER_H
+
+#include <stdexcept>
+#include <string>
+#include <thread>
+#include <atomic>
+#include <functional>
+#include <iostream>
+
+#include <boost/asio.hpp>
+#include <boost/bind.hpp>
+
+#include "psen_scan/open_connection_failure.h"
+#include "psen_scan/close_connection_failure.h"
+
+namespace psen_scan
+{
+class InvalidDataHandler : public std::invalid_argument
+{
+public:
+  InvalidDataHandler(const std::string& msg = "New data handler is invalid");
+};
+
+class InvalidErrorHandler : public std::invalid_argument
+{
+public:
+  InvalidErrorHandler(const std::string& msg = "Error handler is invalid");
+};
+
+template <std::size_t NumberOfBytes>
+using NewDataHandler = std::function<void(const std::array<char, NumberOfBytes>&, const std::size_t&)>;
+
+using ErrorHandler = std::function<void(const std::string&)>;
+
+static const boost::posix_time::seconds DEFAULT_TIMEOUT{ 1 };
+
+template <std::size_t NumberOfBytes>
+class AsynUdpReader
+{
+public:
+  AsynUdpReader(const NewDataHandler<NumberOfBytes>& data_handler,
+                const ErrorHandler& error_handler,
+                const unsigned short& host_port,
+                const std::string& endpoint_ip,
+                const unsigned short& endpoint_port);
+  ~AsynUdpReader();
+
+public:
+  void startReceiving(const boost::posix_time::time_duration timeout = DEFAULT_TIMEOUT);
+  void stopReceiving();
+
+  void close();
+
+private:
+  void asyncReceive(const boost::posix_time::time_duration timeout);
+  void handleTimeout(const boost::system::error_code& error_code);
+  void handleReceive(const boost::system::error_code& error_code,
+                     const std::size_t& bytes_received,
+                     const boost::posix_time::time_duration timeout);
+
+private:
+  boost::asio::io_service io_service_;
+  // Prevent the run() method of the io_service from returning when there is no more work.
+  boost::asio::io_service::work work_{ io_service_ };
+  boost::asio::deadline_timer timeout_timer_{ io_service_ };
+  std::thread io_service_thread_;
+
+  std::array<char, NumberOfBytes> received_data_;
+  NewDataHandler<NumberOfBytes> data_handler_;
+  ErrorHandler error_handler_;
+
+  boost::asio::ip::udp::socket socket_;
+  boost::asio::ip::udp::endpoint endpoint_;
+
+  enum class Receiving
```

Comment:
This introduces quite some complexity. https://www.boost.org/doc/libs/1_35_0/doc/html/boost_asio/tutorial/tutdaytime6/src.html does not seem to need it. Could you provide some explanation?

Reply:
Removed-> Done.